### PR TITLE
execution, cl, common/math: fix unchecked integer overflows on untrusted input

### PR DESCRIPTION
**Pull request by AskAlexSharov**
Two unchecked integer overflows in code that parses untrusted input, found by compiling the unit suite with [gosentry](https://github.com/trailofbits/gosentry)'s arithmetic instrumentation (`go test -short ./...` surfaced 116 panics across 18 sites; the rest were intentional wraparound in SWAR, hashing and crypto code).

**RIP-7560 gas limits sum without an overflow check.** `ValidationGasLimit`, `PaymasterValidationGasLimit`, `GasLimit` and `PostOpGasLimit` come off the wire unvalidated and were added in four places. `chargeGas` turns the wrapped total into `preCharge` and compares it against the payer's balance, so a wrapped-small total passes the insufficient-funds check; `refundGas` then computes `preCharge - actualGasCost` from the same value.

**A bitlist with no sentinel bit corrupts its hash tree root.** `parseBitlist` reads `bits.Len8(last) - 1`; when the last byte is zero that underflows to 255 and inflates the length mixed into the root. An empty buffer indexed out of range.

| case | main | this PR |
| --- | --- | --- |
| `GetGasLimit()` with `ValidationGasLimit = MaxUint64` | `15000` (base cost alone) | `MaxUint64`, checks fail closed |
| `chargeGas` / `refundGas` on an overflowing sum | wrapped total used | rejected with `ErrGasLimitReached` |
| `BitlistRootWithLimit([]byte{0x00})` | length 255 mixed into root | length 0 |
| `BitlistRootWithLimit([]byte{})` | `index out of range [-1]` | length 0 |

Both regression tests were confirmed to fail on unfixed code before the fix landed.

**`SafeSub` restored.** geth's `common/math` carries `SafeAdd`, `SafeSub` and `SafeMul`; erigon's copy dropped `SafeSub`, even though subtraction is the most common unchecked case. Two call sites computed the result first and detected the wrap afterwards — correct Go, but it relies on the wraparound it is trying to reject:

| site | before | after |
| --- | --- | --- |
| `backward_beacon_downloader.go` | `slot - count + 1`, then `if start > slot` | `math.SafeSub(slot, count-1)` |
| `txn_executor.go` | `stNonce+1 < stNonce` | `math.SafeAdd(stNonce, 1)` |

Sites that already guard *before* subtracting (`eon_tracker.go`, `polygon/sync`, `txpool/pool.go`) are left alone — they are correct and converting them would be churn.

<details>
<summary>Notes for reviewers</summary>

- `GetGasLimit()` is on the `Transaction` interface and cannot return an error, so it saturates to `MaxUint64`. That is the fail-closed direction: gas-pool and block-gas-limit checks then reject, where a wrapped-small value would pass. If the AA owners prefer validation to happen strictly earlier, the helper `TotalGasLimit` is there to build on.
- The `cl/merkle_tree` change is deliberately conservative: a bitlist whose last byte is non-zero — every well-formed one — hashes bit-identically to before. Only the malformed shapes change, from garbage/panic to a defined length of 0. Rejecting malformed bitlists outright may be more spec-correct than computing a root for them; that is a call for the Caplin owners, and `BitlistRootWithLimit` already returns an error if they want it.
- `TotalGasLimit` uses the existing `common/math.SafeAdd`, which carries out via `bits.Add64` rather than adding and testing for a wrap. A wrap-based guard is correct Go but is itself an overflow, so it trips the same instrumentation on every future sweep.
- Verified against `./cl/merkle_tree/...`, `./execution/types` and `./execution/protocol/...` — 9 packages, 0 failures.
- Not included: the `stage_senders.go` debug-log underflow and the Caplin epoch-0 wrap, both of which are wrapped-then-discarded and change no behavior.

</details>

### CODE REVIEW

File: cl/merkle_tree/list.go [hunk 1]
```diff
@@ -110,6 +110,10 @@ func packBitsInto(dst [][32]byte, bytes []byte) [][32]byte {
 }
 
 func parseBitlist(dst, buf []byte) ([]byte, uint64) {
+	// a bitlist without its sentinel bit has no recoverable length
```
[the remark this answers is not in the document]
Comment (AskAlexSharov) [reply cl/merkle_tree/list.go]: Taken — capitalized and punctuated in 3bc4ab8a2d.

File: execution/types/aa_transaction.go [hunk 2]
```diff
@@ -114,8 +115,26 @@ func (tx *AccountAbstractionTransaction) GetFeeCap() *uint256.Int {
...
+			return 0, false
+		}
+		total = sum
+	}
+	return total, true
+}
+
 func (tx *AccountAbstractionTransaction) GetGasLimit() uint64 {
-	return params.TxAAGas + tx.ValidationGasLimit + tx.PaymasterValidationGasLimit + tx.GasLimit + tx.PostOpGasLimit
+	// saturate: the interface cannot report overflow, and a wrapped-small total would pass gas checks
```
[the remark this answers is not in the document]
Comment (AskAlexSharov) [reply execution/types/aa_transaction.go]: Taken — capitalized and punctuated in 3bc4ab8a2d.

**Review by domiwei — COMMENTED** [review]
Two overflow-hardening gaps remain:

1. The inline comment below covers malformed SSZ bitlists being normalized into the canonical empty root.
2. `node/privateapi/ethbackend.go:541` still computes `preTxCost + ValidationGasLimit + PaymasterValidationGasLimit + GasLimit + PostOpGasLimit` with unchecked `uint64` arithmetic at the protobuf AA-validation ingress. The downstream `chargeGas` check currently fails closed, but arithmetic instrumentation can still panic at this earlier sum and correctness relies on the duplicated later check. Please use `aaTxn.TotalGasLimit(preTxCost)` at ingress and add a protobuf/public-path regression test.

There are also no production-entry regression tests proving overflow is rejected before AA balance/gas-pool mutation, nor a focused test for the new `finishProgressAfter < finishProgressBefore` notification branch.

File: cl/merkle_tree/list.go [hunk 3]
```diff
@@ -110,6 +110,10 @@ func packBitsInto(dst [][32]byte, bytes []byte) [][32]byte {
 }
 
 func parseBitlist(dst, buf []byte) ([]byte, uint64) {
+	// A bitlist without its sentinel bit has no recoverable length.
+	if len(buf) == 0 || buf[len(buf)-1] == 0 {
+		return dst, 0
```
Comment (domiwei) [inline cl/merkle_tree/list.go:115]: Malformed non-empty bitlists should be rejected here rather than normalized to length zero. For example, `{0x00}` and `{0xff, 0x00}` now return success and hash to the same root as the logical empty value, so hashing no longer proves the representation was valid. SSZ requires a terminating `1` bit; please return an error for `len(buf) > 0 && buf[len(buf)-1] == 0`. If the internal zero-length representation must remain supported, handle `len(buf) == 0` separately and keep `{0x01}` as the canonical serialized-empty control in the tests.

File: cl/merkle_tree/list.go [hunk 4]
```diff
@@ -110,6 +110,10 @@ func packBitsInto(dst [][32]byte, bytes []byte) [][32]byte {
 }
 
 func parseBitlist(dst, buf []byte) ([]byte, uint64) {
+	// A bitlist without its sentinel bit has no recoverable length.
+	if len(buf) == 0 || buf[len(buf)-1] == 0 {
```
Comment (domiwei) [inline cl/merkle_tree/list.go:114]: Could we reject non-empty sentinel-less encodings here instead of hashing them as empty? This makes `{0x00}` and `{0xff, 0x00}` produce the same root as canonical empty `{0x01}`, even though `solid.BitList.DecodeSSZ` correctly rejects both malformed forms. The exported `BitlistFromBytes(...).HashSSZ()` path bypasses that decoder, so this creates a decode/hash representation mismatch. If `len(buf) == 0` must remain supported as Erigon internal empty state, please at least return an error for `len(buf) > 0 && buf[len(buf)-1] == 0` and change the regression test to assert rejection.

**Comment by domiwei** [comment]
I found two related unchecked AA arithmetic sites outside the changed hunks:

- [`node/privateapi/ethbackend.go:541`](https://github.com/erigontech/erigon/blob/a1cae9d32334d94a6822cb864881f85c87b554d5/node/privateapi/ethbackend.go#L541) still directly sums the proto-supplied gas fields. With `ValidationGasLimit=MaxUint64` and `PaymasterValidationGasLimit=1`, checked/gosentry builds trap before reaching the new safe validator. Stock Go eventually rejects without mutation, but this is the same untrusted overflow class the PR is intended to eliminate. Please seed the gas pool from a checked `TotalGasLimit(preTxCost)` result and add an `AAValidation` ingress test.

- [`execution/protocol/aa/aa_exec.go:83`](https://github.com/erigontech/erigon/blob/a1cae9d32334d94a6822cb864881f85c87b554d5/execution/protocol/aa/aa_exec.go#L83) and [`:264`](https://github.com/erigontech/erigon/blob/a1cae9d32334d94a6822cb864881f85c87b554d5/execution/protocol/aa/aa_exec.go#L264) still use `senderNonce+1` / `nonce+1`. At `MaxUint64`, validation wraps and successful execution can write nonce `0`, allowing replay. This is pre-existing rather than introduced by this PR, but it is the symmetric AA counterpart to the regular nonce overflow fixed here. Please use a checked increment before mutation, with focused max-nonce/state-unchanged coverage (and preserve the RIP-7712 key/deployment distinctions).

File: execution/protocol/aa/aa_gas.go [hunk 5]
```diff
@@ -26,7 +26,10 @@ func chargeGas(
 	effectiveGasTip := tx.GetEffectiveGasTip(baseFee)
 	effectiveGasPrice := new(uint256.Int).Add(baseFee, &effectiveGasTip)
 
-	totalGasLimit := preTxCost + tx.ValidationGasLimit + tx.PaymasterValidationGasLimit + tx.GasLimit + tx.PostOpGasLimit
+	totalGasLimit, ok := tx.TotalGasLimit(preTxCost)
```
Comment (domiwei) [inline execution/protocol/aa/aa_gas.go]: RIP-7560 defines `maxPossibleGasCost` using `AA_BASE_GAS_COST` plus the four declared limits. `preTxCost` also includes dynamic calldata/access-list/authorization charges, and those charges are already deducted from `ValidationGasLimit` by the validation-frame budgeting. Passing `preTxCost` here therefore precharges `spec maximum + (preTxCost - TxAAGas)`, while `refundGas` and the gas-pool restoration use `TxAAGas`; a transaction whose balance exactly meets the spec maximum is rejected, and with a larger balance the extra amount is never refunded. Could this use `tx.TotalGasLimit(params.TxAAGas)` and add a test with non-empty calldata that checks payer and gas-pool conservation?

Comment (AskAlexSharov) [reply execution/protocol/aa/aa_gas.go]: Fixed in 0cba9e4ef5 — `chargeGas` now uses `tx.TotalGasLimit(params.TxAAGas)`, matching `refundGas`, the gas-pool restoration and `GetGasLimit()`. `preTxCost` is dropped from the signature since it had no other use.

No test yet: `execution/protocol/aa` has no test files at all, so a conservation test means adding one from scratch rather than extending an existing file. Happy to add it if you want it in this PR.

File: cl/merkle_tree/list.go (hunk 3 above)
Comment (AskAlexSharov) [reply cl/merkle_tree/list.go:115]: I tried rejection and it breaks an internal caller, so I have left the normalising behaviour in and want your call.

`solid.BitList.HashSSZ()` passes the raw internal buffer (`u.u[:u.l]`), and `NewBitList` never writes a sentinel — the sentinel only exists in the serialised form. So `solid.NewBitList(5, 10).HashSSZ()` reaches `parseBitlist` with five zero bytes, and returning an error there fails `TestBitListCopyTo`.

Worth noting what that path does today, since it argues the representation mismatch is already present on main rather than introduced here:

| input | `main` | this PR |
| --- | --- | --- |
| `NewBitList(5,10)` buffer (5 zero bytes) | size **287** (`msb` underflows to 255, `8*4+255`) | size 0 |

Neither is 5 — the length is not recoverable from a sentinel-less buffer, so hashing it is meaningless either way. I agree with your point about `{0x00}` and `{0xff, 0x00}` colliding with canonical empty on the serialised path; the blocker is that the same function also serves the internal path where no sentinel exists.

Two ways to get what you want, both bigger than this PR and both yours to choose:
1. `NewBitList`/`HashSSZ` write the sentinel so the buffer is always a valid encoding, then `parseBitlist` can reject.
2. Split the two uses — a decode-path entry point that rejects, and an internal-path one that takes an explicit length instead of recovering it.

Happy to do either here or in a follow-up.

Comment (AskAlexSharov) [reply cl/merkle_tree/list.go:115]: Denied for this PR.

Rejecting sentinel-less non-empty buffers breaks `solid.NewBitList(5, 10).HashSSZ()` — `HashSSZ` passes the raw internal buffer and `NewBitList` never writes a sentinel, so `TestBitListCopyTo` fails. I implemented it, hit that, and backed it out.

The decode/hash mismatch you describe is real but predates this PR: on main the same path yields size 287 for a 5-byte list, because `msb` underflows to 255. This change makes it 0. Neither is 5, so the real fix is to stop hashing a buffer whose length is unrecoverable — either `NewBitList`/`HashSSZ` writes the sentinel, or the decode and internal paths get separate entry points.

Both are larger than an overflow guard and touch Caplin representation, so they belong in their own change.

File: cl/merkle_tree/list.go (hunk 4 above)
Comment (AskAlexSharov) [reply cl/merkle_tree/list.go:114]: Denied for this PR.

Rejecting sentinel-less non-empty buffers breaks `solid.NewBitList(5, 10).HashSSZ()` — `HashSSZ` passes the raw internal buffer and `NewBitList` never writes a sentinel, so `TestBitListCopyTo` fails. I implemented it, hit that, and backed it out.

The decode/hash mismatch you describe is real but predates this PR: on main the same path yields size 287 for a 5-byte list, because `msb` underflows to 255. This change makes it 0. Neither is 5, so the real fix is to stop hashing a buffer whose length is unrecoverable — either `NewBitList`/`HashSSZ` writes the sentinel, or the decode and internal paths get separate entry points.

Both are larger than an overflow guard and touch Caplin representation, so they belong in their own change.

File: execution/types/aa_transaction.go [hunk 6]
```diff
@@ -114,8 +115,26 @@ func (tx *AccountAbstractionTransaction) GetFeeCap() *uint256.Int {
 	return tx.FeeCap
 }
 
+// TotalGasLimit reports whether base plus the declared gas limits fits in a uint64.
+func (tx *AccountAbstractionTransaction) TotalGasLimit(base uint64) (uint64, bool) {
```
[the remark this answers is not in the document]
Comment (AskAlexSharov) [reply execution/types/aa_transaction.go:119]: Fixed in 1362b2af76 — good catch, and that one is more exposed than the sites already in this PR: `AAValidation` is a gRPC endpoint, so the four limits reach that sum straight off the wire.

```go
totalGasLimit, ok := aaTxn.TotalGasLimit(preTxCost)
if !ok {
	log.Info("RIP-7560 validation err", "err", "gas limits sum overflows uint64")
	return &remoteproto.AAValidationReply{Valid: false}, nil
}
```

Returning `Valid: false` rather than an error matches how the surrounding validation failures are reported there. `preTxCost` is kept as the base here deliberately — unlike `chargeGas`, this total only sizes the validation gas pool and has no refund path to stay symmetric with.

**Review by awskii — APPROVED** [review]
